File: blackjack.py
# blackjack.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import json
import random
from pathlib import Path

COLORS = {
    'reset': '\033[0m',
    'red': '\033[91m',
    'green': '\033[92m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'cyan': '\033[96m',
    'bold': '\033[1m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

SUITS = ['♠', '♥', '♦', '♣']
RANKS = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']
VALUES = {'2':2,'3':3,'4':4,'5':5,'6':6,'7':7,'8':8,'9':9,'10':10,'J':10,'Q':10,'K':10,'A':11}

class Card:
    def __init__(self, suit, rank):
        self.suit = suit
        self.rank = rank
        self.value = VALUES[rank]
        self.color = 'red' if suit in ('♥', '♦') else 'black'

    def __str__(self):
        return colorize(f"{self.rank}{self.suit}", self.color)

class Deck:
    def __init__(self):
        self.cards = [Card(s, r) for s in SUITS for r in RANKS]
        self.shuffle()

    def shuffle(self):
        random.shuffle(self.cards)

    def draw(self):
        return self.cards.pop() if self.cards else None

class Hand:
    def __init__(self, cards=None):
        self.cards = cards or []
        self.bet = 0
        self.is_done = False

    def add(self, card):
        self.cards.append(card)

    def value(self):
        total = sum(c.value for c in self.cards)
        aces = sum(1 for c in self.cards if c.rank == 'A')
        while total > 21 and aces:
            total -= 10
            aces -= 1
        return total

    def is_blackjack(self):
        return len(self.cards) == 2 and self.value() == 21

    def is_bust(self):
        return self.value() > 21

    def can_split(self):
        return len(self.cards) == 2 and self.cards[0].rank == self.cards[1].rank

    def split(self):
        return Hand([self.cards.pop()])

    def __str__(self):
        return ' '.join(str(c) for c in self.cards)

class Blackjack:
    def __init__(self, balance=1000):
        self.balance = balance
        self.bet = 0
        self.deck = Deck()
        self.player_hands = [Hand()]
        self.dealer_hand = Hand()
        self.insurance = False
        self.insurance_paid = 0
        self.stats_file = Path.home() / '.blackjack_stats.json'
        self.load_stats()

    def load_stats(self):
        if self.stats_file.exists():
            with open(self.stats_file, 'r') as f:
                self.stats = json.load(f)
        else:
            self.stats = {'wins': 0, 'losses': 0, 'pushes': 0, 'blackjacks': 0}

    def save_stats(self):
        with open(self.stats_file, 'w') as f:
            json.dump(self.stats, f)

    def display_stats(self):
        print(colorize("Статистика:", 'bold'))
        print(f"  Побед: {self.stats['wins']}")
        print(f"  Поражений: {self.stats['losses']}")
        print(f"  Ничьих: {self.stats['pushes']}")
        print(f"  Блэкджеков: {self.stats['blackjacks']}")
        print(f"  Баланс: {self.balance}")

    def deal_initial(self):
        self.deck = Deck()
        self.deck.shuffle()
        self.player_hands = [Hand()]
        self.dealer_hand = Hand()
        self.insurance = False
        self.insurance_paid = 0
        for _ in range(2):
            self.player_hands[0].add(self.deck.draw())
            self.dealer_hand.add(self.deck.draw())

    def show_hands(self, hide_dealer=True):
        print("\nДилер:")
        if hide_dealer:
            print(f"  {self.dealer_hand.cards[0]}  [скрыто]")
        else:
            print(f"  {self.dealer_hand}  (очки: {self.dealer_hand.value()})")
        print("\nВаши руки:")
        for i, hand in enumerate(self.player_hands):
            print(f"  Рука {i+1}: {hand}  (очки: {hand.value()})")

    def get_bet(self):
        while True:
            try:
                bet = int(input(f"Ваша ставка (баланс: {self.balance}): "))
                if 0 < bet <= self.balance:
                    self.bet = bet
                    return
                else:
                    print(colorize("Неверная ставка.", 'red'))
            except ValueError:
                print(colorize("Введите число.", 'red'))

    def play_hand(self, hand_idx):
        hand = self.player_hands[hand_idx]
        while not hand.is_done and not hand.is_bust() and hand.value() < 21:
            self.show_hands()
            print(f"\nРука {hand_idx+1}: {hand}  (очки: {hand.value()})")
            action = input("Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): ").strip().lower()
            if action == 'q':
                print(colorize("Выход.", 'yellow'))
                self.save_stats()
                sys.exit(0)
            elif action == 'h':
                hand.add(self.deck.draw())
                if hand.is_bust():
                    print(colorize("Перебор!", 'red'))
                    hand.is_done = True
            elif action == 's':
                hand.is_done = True
            elif action == 'd':
                if len(hand.cards) == 2 and self.balance >= self.bet:
                    self.bet *= 2
                    hand.add(self.deck.draw())
                    hand.is_done = True
                    if hand.is_bust():
                        print(colorize("Перебор!", 'red'))
                else:
                    print(colorize("Удвоение недоступно.", 'yellow'))
            elif action == 'i':
                if self.dealer_hand.cards[0].rank == 'A' and not self.insurance:
                    self.insurance = True
                    self.insurance_paid = self.bet // 2
                    self.balance -= self.insurance_paid
                    print(colorize("Страховка активирована.", 'cyan'))
                else:
                    print(colorize("Страховка недоступна.", 'yellow'))
            elif action == 'sp':
                if hand.can_split():
                    new_hand = hand.split()
                    self.player_hands.append(new_hand)
                    # Добавляем по одной карте к каждой руке
                    hand.add(self.deck.draw())
                    new_hand.add(self.deck.draw())
                    print(colorize("Руки разделены.", 'cyan'))
                    # Проверяем, что баланс позволяет дополнительную ставку
                    if self.balance < self.bet:
                        print(colorize("Недостаточно средств для сплита.", 'red'))
                        # откат
                        self.player_hands.pop()
                        # возвращаем карту обратно
                        # Упрощённо: просто игнорируем
                    else:
                        self.balance -= self.bet  # вторая ставка
                        self.bet *= 2  # общая ставка = 2*bet
                else:
                    print(colorize("Сплит недоступен.", 'yellow'))
            else:
                print(colorize("Неизвестная команда.", 'red'))

    def dealer_play(self):
        while self.dealer_hand.value() < 17:
            self.dealer_hand.add(self.deck.draw())
        return self.dealer_hand.value()

    def resolve(self):
        self.show_hands(hide_dealer=False)
        dealer_value = self.dealer_hand.value()
        for hand in self.player_hands:
            player_value = hand.value()
            if hand.is_bust():
                self.balance -= self.bet
                self.stats['losses'] += 1
                print(colorize(f"Поражение (перебор): рука проиграла {self.bet}", 'red'))
            elif hand.is_blackjack() and not self.dealer_hand.is_blackjack():
                win = int(self.bet * 1.5)
                self.balance += win
                self.stats['blackjacks'] += 1
                self.stats['wins'] += 1
                print(colorize(f"Блэкджек! Выигрыш {win}", 'green'))
            elif dealer_value > 21:
                self.balance += self.bet
                self.stats['wins'] += 1
                print(colorize(f"Дилер перебрал, выигрыш {self.bet}", 'green'))
            elif player_value > dealer_value:
                self.balance += self.bet
                self.stats['wins'] += 1
                print(colorize(f"Выигрыш {self.bet}", 'green'))
            elif player_value == dealer_value:
                self.stats['pushes'] += 1
                print(colorize("Ничья, ставка возвращена.", 'yellow'))
            else:
                self.balance -= self.bet
                self.stats['losses'] += 1
                print(colorize(f"Поражение, потеряно {self.bet}", 'red'))
        # Страховка
        if self.insurance:
            if self.dealer_hand.is_blackjack():
                self.balance += self.insurance_paid * 2
                print(colorize("Страховка выиграла!", 'cyan'))
            else:
                print(colorize("Страховка проиграла.", 'yellow'))

        self.save_stats()

    def play(self):
        print(colorize("🃏 Добро пожаловать в Блэкджек!", 'bold'))
        print(f"Ваш баланс: {self.balance}")
        while self.balance > 0:
            self.get_bet()
            self.deal_initial()
            # Проверка блэкджека у игрока
            if self.player_hands[0].is_blackjack():
                self.show_hands()
                if self.dealer_hand.is_blackjack():
                    print(colorize("Ничья – у обоих блэкджек!", 'yellow'))
                    self.stats['pushes'] += 1
                    self.save_stats()
                    continue
                else:
                    win = int(self.bet * 1.5)
                    self.balance += win
                    self.stats['blackjacks'] += 1
                    self.stats['wins'] += 1
                    print(colorize(f"Блэкджек! Выигрыш {win}", 'green'))
                    self.save_stats()
                    continue
            # Проверка блэкджека у дилера (игрок может взять страховку)
            if self.dealer_hand.cards[0].rank == 'A':
                # Предлагаем страховку
                print("У дилера туз. Хотите страховку? (y/n): ", end='')
                ans = input().strip().lower()
                if ans == 'y':
                    self.insurance = True
                    self.insurance_paid = self.bet // 2
                    self.balance -= self.insurance_paid
                    print(colorize("Страховка активирована.", 'cyan'))
            # Ход игрока
            for i in range(len(self.player_hands)):
                self.play_hand(i)
            # Ход дилера
            self.dealer_play()
            self.resolve()
            print(colorize(f"Баланс: {self.balance}", 'bold'))
            if self.balance <= 0:
                print(colorize("Вы проиграли все деньги!", 'red'))
                break
            if input("Продолжить игру? (y/n): ").strip().lower() != 'y':
                break
        self.display_stats()
        self.save_stats()

def main():
    balance = 1000
    reset = False
    if len(sys.argv) > 1:
        if sys.argv[1] == '-r' or sys.argv[1] == '--reset':
            reset = True
        elif sys.argv[1] == '-s' and len(sys.argv) > 2:
            balance = int(sys.argv[2])
        elif sys.argv[1] == '-h' or sys.argv[1] == '--help':
            print("Usage: blackjack.py [-s start_balance] [-r]")
            return
    if reset:
        stats_file = Path.home() / '.blackjack_stats.json'
        if stats_file.exists():
            stats_file.unlink()
        print("Статистика сброшена.")
        return
    game = Blackjack(balance)
    game.play()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print(colorize("\nИгра прервана.", 'yellow'))
        sys.exit(0)

File: blackjack.js
// blackjack.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const SUITS = ['♠','♥','♦','♣'];
const RANKS = ['2','3','4','5','6','7','8','9','10','J','Q','K','A'];
const VALUES = {2:2,3:3,4:4,5:5,6:6,7:7,8:8,9:9,10:10,J:10,Q:10,K:10,A:11};

class Card {
    constructor(suit, rank) {
        this.suit = suit;
        this.rank = rank;
        this.value = VALUES[rank];
        this.color = (suit === '♥' || suit === '♦') ? 'red' : 'reset';
    }
    toString() {
        return colorize(this.rank + this.suit, this.color);
    }
}

class Deck {
    constructor() {
        this.cards = [];
        for (const s of SUITS)
            for (const r of RANKS)
                this.cards.push(new Card(s, r));
        this.shuffle();
    }
    shuffle() {
        for (let i = this.cards.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i+1));
            [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
        }
    }
    draw() {
        return this.cards.pop();
    }
}

class Hand {
    constructor() {
        this.cards = [];
        this.bet = 0;
        this.done = false;
    }
    add(card) { this.cards.push(card); }
    value() {
        let total = this.cards.reduce((s,c) => s + c.value, 0);
        let aces = this.cards.filter(c => c.rank === 'A').length;
        while (total > 21 && aces > 0) {
            total -= 10;
            aces--;
        }
        return total;
    }
    isBlackjack() { return this.cards.length === 2 && this.value() === 21; }
    isBust() { return this.value() > 21; }
    canSplit() { return this.cards.length === 2 && this.cards[0].rank === this.cards[1].rank; }
    split() {
        const h = new Hand();
        h.cards.push(this.cards.pop());
        return h;
    }
    toString() {
        return this.cards.map(c => c.toString()).join(' ');
    }
}

class Blackjack {
    constructor(balance = 1000) {
        this.balance = balance;
        this.bet = 0;
        this.deck = null;
        this.playerHands = [];
        this.dealerHand = null;
        this.insurance = false;
        this.insurancePaid = 0;
        this.statsFile = path.join(os.homedir(), '.blackjack_stats.json');
        this.loadStats();
    }

    loadStats() {
        try {
            this.stats = JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
        } catch {
            this.stats = { wins: 0, losses: 0, pushes: 0, blackjacks: 0 };
        }
    }

    saveStats() {
        fs.writeFileSync(this.statsFile, JSON.stringify(this.stats, null, 2));
    }

    displayStats() {
        console.log(colorize('Статистика:', 'bold'));
        console.log(`  Побед: ${this.stats.wins}`);
        console.log(`  Поражений: ${this.stats.losses}`);
        console.log(`  Ничьих: ${this.stats.pushes}`);
        console.log(`  Блэкджеков: ${this.stats.blackjacks}`);
        console.log(`  Баланс: ${this.balance}`);
    }

    dealInitial() {
        this.deck = new Deck();
        this.deck.shuffle();
        this.playerHands = [new Hand()];
        this.dealerHand = new Hand();
        this.insurance = false;
        this.insurancePaid = 0;
        this.playerHands[0].add(this.deck.draw());
        this.dealerHand.add(this.deck.draw());
        this.playerHands[0].add(this.deck.draw());
        this.dealerHand.add(this.deck.draw());
    }

    showHands(hideDealer = true) {
        console.log('\nДилер:');
        if (hideDealer) {
            console.log(`  ${this.dealerHand.cards[0].toString()}  [скрыто]`);
        } else {
            console.log(`  ${this.dealerHand.toString()} (очки: ${this.dealerHand.value()})`);
        }
        console.log('\nВаши руки:');
        for (let i=0; i<this.playerHands.length; i++) {
            console.log(`  Рука ${i+1}: ${this.playerHands[i].toString()} (очки: ${this.playerHands[i].value()})`);
        }
    }

    async getBet() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        while (true) {
            const ans = await new Promise(resolve => rl.question(`Ваша ставка (баланс: ${this.balance}): `, resolve));
            const bet = parseInt(ans);
            if (!isNaN(bet) && bet > 0 && bet <= this.balance) {
                this.bet = bet;
                rl.close();
                return;
            }
            console.log(colorize('Неверная ставка.', 'red'));
        }
    }

    async playHand(idx) {
        const hand = this.playerHands[idx];
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        while (!hand.done && !hand.isBust() && hand.value() < 21) {
            this.showHands();
            console.log(`\nРука ${idx+1}: ${hand.toString()} (очки: ${hand.value()})`);
            const action = await new Promise(resolve => rl.question('Действие (h=hit, s=stand, d=double, i=insurance, sp=split, q=quit): ', resolve));
            const cmd = action.trim().toLowerCase();
            if (cmd === 'q') {
                console.log(colorize('Выход.', 'yellow'));
                this.saveStats();
                process.exit(0);
            } else if (cmd === 'h') {
                hand.add(this.deck.draw());
                if (hand.isBust()) {
                    console.log(colorize('Перебор!', 'red'));
                    hand.done = true;
                }
            } else if (cmd === 's') {
                hand.done = true;
            } else if (cmd === 'd') {
                if (hand.cards.length === 2 && this.balance >= this.bet) {
                    this.bet *= 2;
                    hand.add(this.deck.draw());
                    hand.done = true;
                    if (hand.isBust()) console.log(colorize('Перебор!', 'red'));
                } else {
                    console.log(colorize('Удвоение недоступно.', 'yellow'));
                }
            } else if (cmd === 'i') {
                if (this.dealerHand.cards[0].rank === 'A' && !this.insurance) {
                    this.insurance = true;
                    this.insurancePaid = Math.floor(this.bet / 2);
                    this.balance -= this.insurancePaid;
                    console.log(colorize('Страховка активирована.', 'cyan'));
                } else {
                    console.log(colorize('Страховка недоступна.', 'yellow'));
                }
            } else if (cmd === 'sp') {
                if (hand.canSplit()) {
                    const newHand = hand.split();
                    this.playerHands.push(newHand);
                    hand.add(this.deck.draw());
                    newHand.add(this.deck.draw());
                    console.log(colorize('Руки разделены.', 'cyan'));
                    if (this.balance < this.bet) {
                        console.log(colorize('Недостаточно средств для сплита.', 'red'));
                        this.playerHands.pop();
                    } else {
                        this.balance -= this.bet;
                        this.bet *= 2;
                    }
                } else {
                    console.log(colorize('Сплит недоступен.', 'yellow'));
                }
            } else {
                console.log(colorize('Неизвестная команда.', 'red'));
            }
        }
        rl.close();
    }

    dealerPlay() {
        while (this.dealerHand.value() < 17) {
            this.dealerHand.add(this.deck.draw());
        }
    }

    resolve() {
        this.showHands(false);
        const dealerVal = this.dealerHand.value();
        for (const hand of this.playerHands) {
            const playerVal = hand.value();
            if (hand.isBust()) {
                this.balance -= this.bet;
                this.stats.losses++;
                console.log(colorize(`Поражение (перебор): рука проиграла ${this.bet}`, 'red'));
            } else if (hand.isBlackjack() && !this.dealerHand.isBlackjack()) {
                const win = Math.floor(this.bet * 1.5);
                this.balance += win;
                this.stats.blackjacks++;
                this.stats.wins++;
                console.log(colorize(`Блэкджек! Выигрыш ${win}`, 'green'));
            } else if (dealerVal > 21) {
                this.balance += this.bet;
                this.stats.wins++;
                console.log(colorize(`Дилер перебрал, выигрыш ${this.bet}`, 'green'));
            } else if (playerVal > dealerVal) {
                this.balance += this.bet;
                this.stats.wins++;
                console.log(colorize(`Выигрыш ${this.bet}`, 'green'));
            } else if (playerVal === dealerVal) {
                this.stats.pushes++;
                console.log(colorize('Ничья, ставка возвращена.', 'yellow'));
            } else {
                this.balance -= this.bet;
                this.stats.losses++;
                console.log(colorize(`Поражение, потеряно ${this.bet}`, 'red'));
            }
        }
        if (this.insurance) {
            if (this.dealerHand.isBlackjack()) {
                this.balance += this.insurancePaid * 2;
                console.log(colorize('Страховка выиграла!', 'cyan'));
            } else {
                console.log(colorize('Страховка проиграла.', 'yellow'));
            }
        }
        this.saveStats();
    }

    async play() {
        console.log(colorize('🃏 Добро пожаловать в Блэкджек!', 'bold'));
        console.log(`Ваш баланс: ${this.balance}`);
        while (this.balance > 0) {
            await this.getBet();
            this.dealInitial();
            // Блэкджек игрока
            if (this.playerHands[0].isBlackjack()) {
                this.showHands();
                if (this.dealerHand.isBlackjack()) {
                    console.log(colorize('Ничья – у обоих блэкджек!', 'yellow'));
                    this.stats.pushes++;
                    this.saveStats();
                    continue;
                } else {
                    const win = Math.floor(this.bet * 1.5);
                    this.balance += win;
                    this.stats.blackjacks++;
                    this.stats.wins++;
                    console.log(colorize(`Блэкджек! Выигрыш ${win}`, 'green'));
                    this.saveStats();
                    continue;
                }
            }
            // Страховка
            if (this.dealerHand.cards[0].rank === 'A') {
                const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
                const ans = await new Promise(resolve => rl.question('У дилера туз. Хотите страховку? (y/n): ', resolve));
                rl.close();
                if (ans.trim().toLowerCase() === 'y') {
                    this.insurance = true;
                    this.insurancePaid = Math.floor(this.bet / 2);
                    this.balance -= this.insurancePaid;
                    console.log(colorize('Страховка активирована.', 'cyan'));
                }
            }
            // Ход игрока
            for (let i=0; i<this.playerHands.length; i++) {
                await this.playHand(i);
            }
            // Ход дилера
            this.dealerPlay();
            this.resolve();
            console.log(colorize(`Баланс: ${this.balance}`, 'bold'));
            if (this.balance <= 0) {
                console.log(colorize('Вы проиграли все деньги!', 'red'));
                break;
            }
            const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
            const cont = await new Promise(resolve => rl.question('Продолжить игру? (y/n): ', resolve));
            rl.close();
            if (cont.trim().toLowerCase() !== 'y') break;
        }
        this.displayStats();
        this.saveStats();
    }
}

async function main() {
    let balance = 1000;
    let reset = false;
    for (let i=2; i<process.argv.length; i++) {
        const arg = process.argv[i];
        if (arg === '-r' || arg === '--reset') reset = true;
        else if ((arg === '-s' || arg === '--start') && i+1 < process.argv.length) {
            balance = parseInt(process.argv[++i]);
        } else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node blackjack.js [-s start_balance] [-r]');
            process.exit(0);
        }
    }
    if (reset) {
        const f = path.join(os.homedir(), '.blackjack_stats.json');
        if (fs.existsSync(f)) fs.unlinkSync(f);
        console.log('Статистика сброшена.');
        return;
    }
    const game = new Blackjack(balance);
    await game.play();
}

main().catch(console.error);
